### PR TITLE
Add support for clangd's semantic highlighting

### PR DESCRIPTION
**Pull request by DavidGoldman**
- Add LSP types for semantic highlighting
- Limited to clients which support dynamic registration for semantic
  highlighting
- Requires clangd 11 or later

### CODE REVIEW

**Comment by DavidGoldman** [comment]
@swift-ci please test

**Comment by DavidGoldman** [comment]
Before: 
No semantic highlighting


After: 
With semantic highlighting

**Comment by DavidGoldman** [comment]
Also added a warning for the PR comment that I missed on https://github.com/apple/sourcekit-lsp/pull/380

**Comment by akyrtzi** [comment]
Awesome! 🥳

**Comment by DavidGoldman** [comment]
@swift-ci please test

File: Sources/LanguageServerProtocol/Requests/DocumentSemanticTokensDeltaRequest.swift [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+  public init(resultId: String? = nil, edits: [SemanticTokensEdit]) {
+    self.resultId = resultId
+    self.edits = edits
+  }
+}
+
+public struct SemanticTokensEdit: Codable, Hashable {
+  /// Start offset of the edit.
+  public var start: UInt
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/DocumentSemanticTokensDeltaRequest.swift:74]: I think this should use `Int` to better match the conventions of Swift.  This is effectively an array index, which in Swift uses `Int`.  Same for `deleteCount`, which is more like an array length, but the argument is the same.  This also matches everywhere else in SourceKit-LSP where we've used Int.

File: Sources/LanguageServerProtocol/Requests/DocumentSemanticTokensRequest.swift [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  }
+}
+
+public struct DocumentSemanticTokensResponse: ResponseType, Hashable {
+  /// An optional result identifier which enables supporting clients to request semantic token deltas
+  /// subsequent requests.
+  public var resultId: String?
+
+  /// Raw tokens data.
+  public var data: [UInt]
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/Requests/DocumentSemanticTokensRequest.swift:31]: Same comment about `UInt` vs `UInt32` here.

File: Sources/LanguageServerProtocol/SupportTypes/ClientCapabilities.swift [hunk 3]
```diff
@@ -394,6 +424,80 @@ public struct TextDocumentClientCapabilities: Hashable, Codable {
...
+    public var multilineTokenSupport: Bool? = nil
+
+    public init(
+      dynamicRegistration: Bool? = nil,
+      requests: SemanticTokensRequestsClientCapabilities,
+      tokenTypes: [String],
+      tokenModifiers: [String],
+      formats: [TokenFormat],
+      overlappingTokenSupport: Bool? = nil,
+      multilineTokenSupport: Bool? = nil) {
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/SupportTypes/ClientCapabilities.swift:490]: Formatting: `{` or `) {` on a new line.

File: Sources/LanguageServerProtocol/SupportTypes/ServerCapabilities.swift [hunk 4]
```diff
@@ -458,6 +464,41 @@ public struct DocumentLinkOptions: Codable, Hashable {
   }
 }
 
+public struct SemanticTokensRangeOptions: Equatable, Hashable, Codable {
+  // Empty in the LSP 3.16 spec.
+}
+
+public struct SemanticTokensFullOptions: Equatable, Hashable, Codable {
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/SupportTypes/ServerCapabilities.swift:471]: Maybe this (and `SemanticTokensRangeOptions`) should be nested inside `SemanticTokensOptions`?  They're not really useful as top-level types.

File: Sources/LanguageServerProtocol/SupportTypes/SemanticTokens.swift [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+/// The legend for a server's encoding of semantic tokens.
+public struct SemanticTokensLegend: Codable, Hashable {
```
Comment (benlangmuir) [inline Sources/LanguageServerProtocol/SupportTypes/SemanticTokens.swift:14]: This type should have a public `init`.

File: Tests/SourceKitLSPTests/LocalClangTests.swift [hunk 6]
```diff
@@ -184,4 +184,33 @@ final class LocalClangTests: XCTestCase {
       waitForExpectations(timeout: 15)
     }
   }
+
+  func testSemanticHighlighting() throws {
```
Comment (benlangmuir) [inline Tests/SourceKitLSPTests/LocalClangTests.swift:188]: Please update Linux tests.

**Comment by DavidGoldman** [comment]
@swift-ci please test

**Review by benlangmuir — APPROVED** [review]

**Pull request merged**